WIP: struggling to loop through a list of paths and pick first readable.

berkes opened the pull request:
Being new to Rust, I'm struggling to implement a solution for  #18 ( Ignore URL arguments when a file without the arguments exists.)

So far, I managed to refactor the logic that reads the file a bit into the following logic:

1. build a list of candidates for files.
1. loop through that list, the first that can be read is the file to be served.
1. serve that file.

But I'm stuck "fighting" the rust compiler. I am not able to store the contents of the file when read in step 2, to serve it later on.

Before I poor much more time into this, I'd like to know if this direction is acceptable at all. And if so, if you like what I did with the FileProxy struct and so. As is probably clear from my code, I'm new to Rust and come from a OOP background; the code is probably a bit too OOP?

And if this is the right direction: any hints to what I'm doing wrong structure- or logic wise would be very welcome 😄 
```
error[E0381]: use of possibly-uninitialized variable: `file_contents`
  --> devserver_lib/src/lib.rs:77:36
   |
77 |     if let Ok(mut file_contents) = file_contents {
   |                                    ^^^^^^^^^^^^^ use of possibly-uninitialized `file_contents`

```

kettle11 commented:
Thanks for looking into this!

Rust can be a bit tricky at first, but after reading through your code it looks like it's almost working!

`FileProxy looks fine to me! Certainly not too object-oriented for my tastes.

---
```rust
error[E0381]: use of possibly-uninitialized variable: `file_contents`
  --> devserver_lib/src/lib.rs:77:36
   |
77 |     if let Ok(mut file_contents) = file_contents {
   |                                    ^^^^^^^^^^^^^ use of possibly-uninitialized `file_contents`
```

This error can be resolved by initializing `file_contents` to an `std::io::Error` like so:

``` rust
let mut file_contents = Err(Error::new(ErrorKind::NotFound, "No matching file found"));
```

Which makes sense, we're initializing with an Error but if we find a matching file we replace the error with an `Ok` value.

I'll add more comments on the code itself.

kettle11 commented on devserver_lib/src/lib.rs:
```diff
@@ -17,6 +16,26 @@ use std::thread;
+        FileProxy { path: String::from(""), extension: None }
+    }
+    fn read(&self, root_path: &str) -> Result<Vec<u8>> {
+        fs::read(Path::new(root_path).join(&self.path))
```
Probably you need to use the extension here as well, right?

berkes replied:
I'm not entirely sure what you mean, but I do see some issues in my current implementation, regarding the fallback index.html. We might mean the same?

This method simply attempts to read the file at path. which is one of the candidates. Joined with the root_path.

kettle11 commented on devserver_lib/src/lib.rs:
```diff
@@ -17,6 +16,26 @@ use std::thread;
+            Some(ext) => Some(ext),
+            None => None,
+        };
+        FileProxy { path: path, extension: owned_ext }
```
Based on the code you have here the match statement isn't doing anything.

This is a bit weird, but in Rust if the name of a variable is the same as the thing you're assigning in the struct you can initialize the struct like this:

```rust
FileProxy { path, extension }
 ```
instead of:

```rust
FileProxy { path: path, extension: extension }
```

So in this case the function can be reduced to:
```rust
fn new(path: String, extension: Option<String>) -> Self {
    FileProxy {path, extension} // This can also be written as `Self {path, extension}`
}
```

kettle11 commented on devserver_lib/src/lib.rs:
```diff
@@ -44,46 +63,28 @@ fn handle_client<T: Read + Write>(mut stream: T, root_path: &str, reload: bool,
+    let candidates = file_candidates(path);
+    let blank = FileProxy::blank();
+    let chosen = candidates.iter().find(|candidate| {
+        if let Ok(file_contents) = candidate.read(root_path) {
```
Based on your usage I think you have a small misconception about how this line is working, which makes sense because it's weird syntax.

```rust
if let Ok(file_contents) = candidate.read(root_path) {
```
is declaring a new variable called `file_contents` whose scope limited to this if statement. It's **not** assigning to the existing `file_contents`. It's shadowing the existing variable.

So what you want is more like this:
```rust
if let Ok(new_file_contents) = candidate.read(root_path) {
   file_contents = new_file_contents;
   true
}
...
```

kettle11 commented on devserver_lib/src/lib.rs:
```diff
@@ -44,46 +63,28 @@ fn handle_client<T: Read + Write>(mut stream: T, root_path: &str, reload: bool,
         #[cfg(feature = "reload")]
         {
-            if extension == Some("html") && reload {
+            if content_type == "text/html" && reload {
```
I like this change! It's more clearly matches to what's going on.

kettle11 commented on devserver_lib/src/lib.rs:
```diff
@@ -17,6 +16,26 @@ use std::thread;
+        FileProxy { path: path, extension: owned_ext }
+    }
+    fn blank() -> Self {
+        FileProxy { path: String::from(""), extension: None }
```
`String::from("")` and `String::new()` do the same thing.

`String::from("")` might do *slightly* more work behind the scenes, but maybe is slightly clearer. It's a matter of personal preference, but I thought I'd point it out.

kettle11 commented on devserver_lib/src/lib.rs:
```diff
@@ -167,6 +168,39 @@ pub fn run(address: &str, port: u32, path: &str, reload: bool, headers: &str) {
     }
 }
 
+fn file_candidates(path: &str) -> Vec<FileProxy> {
```
This function looks good to me!

berkes commented:
I've implemented all your suggestions and -indeed- the missing part how to assign the file_contents works like a charm!

Some notes and question:

> So in this case the function can be reduced to:

I'm not sure if the `::new` now makes a lot of sense. I've left it in, to keep the candidates method a bit cleaner, but the benefit is not that big. Let me know if you want me to remove it.

> String::from("") and String::new() do the same thing.

Eventhough I knew, both were possible, I forgot about it. Thanks for the reminder!

And one larger issue. Not so much with code, but about the logic:

The way I've implemented it, when you have:
```
├── index.html : content: Hello World
└── mars
    └── index.html : content Hello Mars
```

And you request the following, you get the following:

```
localhost:8080/ -> Hello World
localhost:8080/index.html -> Hello World.
localhost:8080/mars/ -> Hello Mars
```

But when you request the following, you unexpectedly get:

```
localhost:8080/venus/ -> Hello World
localhost:8080/app.css -> Hello World
```

All due to the fallback to `index.html`. Which may not always be what you want. What do you think? Should all the latter return a 404? I think so, but thought I'd check first.

berkes commented:
I guess this can be closed as it has long been fixed in f7ae298
